crates/router/src/core/payments/helpers.rs

```diff
@@ -1889,3 +1892,110 @@ impl AttemptType {
+        })
+        .transpose()?;
+    let order_details_metadata_db = parsed_metadata_db
+        .clone()
+        .and_then(|meta| meta.order_details);
+    let order_details_outside_metadata_db = payment_intent.clone().order_details;
+    let order_details_outside_metadata_req = request.clone().order_details;
+    let order_details_metadata_req = request.clone().metadata.and_then(|meta| meta.order_details);
+
+    if order_details_outside_metadata_req
+        .clone()
+        .zip(order_details_metadata_req.clone())
+        .is_some()
+    {
+        Err(errors::ApiErrorResponse::NotSupported { message: "order_details cannot be present both inside and outside metadata in payments request".to_string() })?
+    }
+    if order_details_metadata_db
+        .clone()
+        .zip(order_details_outside_metadata_db.clone())
+        .is_some()
+    {
+        Err(errors::ApiErrorResponse::NotSupported { message: "order_details cannot be present both inside and outside metadata in payment intent in db".to_string() })?
+    }
+    let order_details_outside = match order_details_outside_metadata_req {
+        Some(order) => match order_details_metadata_db {
+            Some(_) => Err(errors::ApiErrorResponse::NotSupported {
+                message: "order_details previously present inside of metadata".to_string(),
+            })?,
+            None => Some(order),
+        },
+        None => match order_details_metadata_req {
+            Some(_order) => match order_details_outside_metadata_db {
+                Some(_) => Err(errors::ApiErrorResponse::NotSupported {
+                    message: "order_details previously present outside of metadata".to_string(),
+                })?,
+                None => None,
+            },
+            None => None,
+        },
+    };
+    add_order_details_and_metadata_to_payment_intent(
+        payment_intent,
+        request,
+        parsed_metadata_db,
+        &order_details_outside,
+    )
+}
+
+pub fn add_order_details_and_metadata_to_payment_intent(
+    mut payment_intent: &mut storage::payment_intent::PaymentIntent,
+    request: &api::PaymentsRequest,
+    parsed_metadata_db: Option<api_models::payments::Metadata>,
+    order_details_outside: &Option<Vec<api_models::payments::OrderDetailsWithAmount>>,
+) -> RouterResult<()> {
+    let metadata_with_order_details = request.clone().metadata.map(|meta| {
+        let transformed_metadata = match parsed_metadata_db {
+            Some(meta_db) => api_models::payments::Metadata {
+                order_details: meta.order_details,
+                ..meta_db
+            },
+            None => meta,
+        };
+        let transformed_metadata_value =
+            Encode::<api_models::payments::Metadata>::encode_to_value(&transformed_metadata)
+                .change_context(errors::ApiErrorResponse::InternalServerError)
+                .attach_printable("Encoding Metadata to value failed")
+                .unwrap_or_default();
```

Comment:
raise the appropriate error instead of doing `unwrap_or_default()`

Reply:
let metadata_with_order_details = match request.clone().metadata{
        Some(meta) => {
            let transformed_metadata = match parsed_metadata_db {
                Some(meta_db) => api_models::payments::Metadata {
                    order_details: meta.order_details,
                    ..meta_db
                },
                None => meta,
            };
            let transformed_metadata_value =
                Encode::<api_models::payments::Metadata>::encode_to_value(&transformed_metadata)
                    .change_context(errors::ApiErrorResponse::InternalServerError)
                    .attach_printable("Encoding Metadata to value failed")?;
            Some(masking::Secret::new(transformed_metadata_value))
        },
        None => None,
    };